src/osd/OSD.cc

```diff
@@ -11534,6 +11494,187 @@ void OSD::ShardedOpWQ::stop_for_fast_shutdown()
   }
 }
 
+// =============================================================
+
+#undef dout_context
+#define dout_context cct
+#undef dout_prefix
+#define dout_prefix *_dout << "OSDBenchTest: "
+
+int OSDBenchTest::precheck()
+{
+  int ret = 0;
+
+  if (!store) {
+    derr << "OSDBenchTest: objectstore not specified!" << dendl;
+    ret = -ENOENT;
+  }
+
+  if (!ch) {
+    derr << "OSDBenchTest: meta collection not specified!" << dendl;
+    ret = -ENOENT;
+  }
+
+  if (!count || !bsize) {
+    ret = -EINVAL;
+  }
+
+  return ret;
+}
+
+int OSDBenchTest::run_test()
+{
+  int ret = 0;
+
+  ret = precheck();
+  if (ret != 0) {
+    return ret;
+  }
+
+  // flush store cache
+  ret = flush_store_cache();
+  if (ret != 0) {
+    return ret;
+  }
+
+  // Prefill
+  prefill_objects();
+
+  // randwrite test
+  randwrite();
+
+  // cleanup
+  cleanup();
+
+  // Calculate bandwidth & iops
+  if (elapsed && bsize) {
+    bandwidth = count / elapsed;
+    iops = bandwidth / bsize;
+    dout(0) << __func__
+            << ": osd bench result -"
+            << std::fixed << std::setprecision(2)
+            << " bandwidth: " << byte_u_t(bandwidth) << "/sec"
+            << " iops: " << iops
+            << " elapsed_sec: " << elapsed << " secs"
+            << dendl;
+
+  } else {
+    ret = -1; // test did not run properly
+  }
+
+  return ret;
+}
+
+void OSDBenchTest::wait_for_flush_commit()
+{
+  C_SaferCond waiter;
+  if (!ch->flush_commit(&waiter)) {
+    waiter.wait();
+  }
+}
+
+void OSDBenchTest::prefill_objects()
+{
+  utime_t start = ceph_clock_now();
+  if (osize && onum) {
```

Comment:
Is there a scenario in which any of these is 0? please document

Reply:
I have added documentation where applicable to make things clearer.